### PR TITLE
Refresh client's grant token if it has expired

### PR DESCRIPTION
**Pull request by robertknight**
During an LTI launch the LMS backend generates a grant token for use by the
Hypothesis client that is valid for 5 minutes (see `JSConfig._grant_token`).

If the Hypothesis client attempts to fetch its configuration after this
time has elapsed then the client will fail to log in. With the sidebar
application this rare because the sidebar launches as soon as Via loads.
It could happen however if the LMS frontend did not load the assignment
content immediately but instead paused for Canvas OAuth and the user
took more than 5 minutes to respond. With the client's "Notebook"
application this is expected to be much more common however, because the
notebook isn't loaded until the user explicitly chooses the "Open
notebook" menu option. That might not happen for minutes or even hours.

Fix the issue by adding a new backend API view, `/api/grant_token` that returns
a new grant token and calling it from the LMS frontend if needed before
responding to a request from the Hypothesis client (which could be from
the sidebar or notebook) for configuration. This view returns a fresh grant token
which is used to update the one in the client config that was initially generated
during the LTI launch.

 - Add `/api/grant_token` route in `lms.views.api.grant_token`
 - Modify `ClientRpc` to check expiry time of grant token and refresh it
   if needed before responding to a `requestConfig` RPC call from the
   client

Part of https://github.com/hypothesis/client/issues/2801
Related [Slack thread](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1615374523004200)

----

**TODO:**

- [ ] Get consensus on the general approach
- [ ] Agree on the route name/path for the backend API view (currently `/api/grant_token`)
- [ ] De-duplicate the logic between `JSConfig._grant_token` and `grant_token`?
- [ ] Write tests for backend and frontend changes

### CODE REVIEW

File: lms/routes.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ def includeme(config):
     )
     config.add_route("canvas_api.sync", "/api/canvas/sync", request_method="POST")
 
+    config.add_route("api.grant_token", "/api/grant_token", request_method="GET")
```
Comment (robertknight) [inline lms/routes.py:67]: Does anyone have recommendations regarding the API name and path or is this good?

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 2]
```diff
@@ -24,17 +59,37 @@ export class ClientRpc {
...
     this._server = new Server(allowedOrigins);
 
-    // Handle the initial request for configuration from the Hypothesis client.
-    this._server.register('requestConfig', () => clientConfig);
+    // Handle the requests for configuration from the Hypothesis client.
+    this._server.register('requestConfig', async () => {
+      // Refresh the grant token used by the client to access the H API, if
+      // it expired.
+      const grantToken = clientConfig.services[0].grantToken;
+      const now = Date.now();
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:79]: Comparing `Date.now()` against the JWT token's "exp" field means there is an assumption that the user's clock is roughly correct. The use of SSL limits how far adrift the user's clock can be and still have the service work, but I don't know how tight that time window is. If the user's clock is ahead of the server we might end up refreshing it when we don't need to - that's OK. If the user's clock is behind it could fail to be refreshed when it needs to be. One way to guard against this would be to just refresh it regardless if the last refresh time was > N minutes ago, according to the LMS frontend.

File: lms/views/api/grant_token.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import jwt
+from pyramid.view import view_config
+
+
+@view_config(permission="api", renderer="json", route_name="api.grant_token")
+def grant_token(request):
+    """
+    Return a grant token that the Hypothesis client can use to log in to H.
+    """
```
Comment (robertknight) [inline lms/views/api/grant_token.py:12]: The code in this function is largely copied and tweaked from `JSConfig._grant_token`. Is there a good shared place to put this logic?

File: lms/routes.py (hunk 1 above)
Comment (lyzadanger) [reply lms/routes.py:67]: I think this seems correct as the resource you are trying to retrieve is a grant token and it's a GET request here...seems about right to me.

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 4]
```diff
@@ -1,11 +1,46 @@
...
+
+/**
+ * Return the time at which a JWT token expires.
+ *
+ * @param {string} jwtToken
+ * @return {number} - JWT expiry timestamp in milliseconds. This can be compared
+ *   with `Date.now()`.
+ */
+function jwtExpiry(jwtToken) {
+  const [, payloadBase64] = jwtToken.split('.');
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:35]: These reference names are nice and self-documenting 👍🏻

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 5]
```diff
@@ -24,17 +59,37 @@ export class ClientRpc {
...
+  constructor({ allowedOrigins, authToken, clientConfig }) {
     this._server = new Server(allowedOrigins);
 
-    // Handle the initial request for configuration from the Hypothesis client.
-    this._server.register('requestConfig', () => clientConfig);
+    // Handle the requests for configuration from the Hypothesis client.
+    this._server.register('requestConfig', async () => {
+      // Refresh the grant token used by the client to access the H API, if
+      // it expired.
+      const grantToken = clientConfig.services[0].grantToken;
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:78]: 0 here feels like a magic number...

File: lms/static/scripts/frontend_apps/services/client-rpc.js (hunk 2 above)
Comment (lyzadanger) [reply lms/static/scripts/frontend_apps/services/client-rpc.js:79]: I think it could be worthwhile to add a brief comment in the code here for future devs.

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 6]
```diff
@@ -1,11 +1,46 @@
...
+ * and not touched by `ClientRpc`.
+ *
+ * See https://h.readthedocs.io/projects/client/en/latest/publishers/config/.
+ *
+ * @typedef ClientConfig
+ * @prop {[ServiceConfig]} services
+ */
+
+/**
+ * Return the time at which a JWT token expires.
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:28]: Orthogonal question: What's our current convention on documenting functions that can `throw`?

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 7]
```diff
@@ -24,17 +59,37 @@ export class ClientRpc {
...
+      // Refresh the grant token used by the client to access the H API, if
+      // it expired.
+      const grantToken = clientConfig.services[0].grantToken;
+      const now = Date.now();
+      if (now > jwtExpiry(grantToken)) {
+        const response = await apiCall({
+          authToken,
+          path: '/api/grant_token',
+        });
+        if (typeof response.grant_token !== 'string') {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:85]: Can you help me understand in what situations there could be a response body but the `grant_token` is non-string?

File: lms/static/scripts/frontend_apps/services/client-rpc.js [hunk 8]
```diff
@@ -24,17 +59,37 @@ export class ClientRpc {
...
+          throw new Error('Invalid grant_token response');
+        }
+        clientConfig.services[0].grantToken = response.grant_token;
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/services/client-rpc.js:88]: Just an observation: we're mutating the config here.

**Review by lyzadanger — COMMENTED** [review]
A few comments on the front-end code. Though the flow control vis-a-vis RPC and launch can be hard to follow at times (nothing new in this code that exacerbates that!), these changes make general sense to me.

File: lms/routes.py (hunk 1 above)
Comment (seanh) [reply lms/routes.py:67]: Yeah, this seems fine. It's a bit ugly perhaps (would `grant-token` be nicer than `grant_token` in a URL?). And should it be under `/api/auth/`? Is "grant token" the best name? That does't really get across what this is: a token for the client to use to log in to h.

But ultimately this is a private API only used by LMS's own frontend code, so URL design probably doesn't matter that much, and is easy to change

File: lms/static/scripts/frontend_apps/services/client-rpc.js (hunk 2 above)
Comment (seanh) [reply lms/static/scripts/frontend_apps/services/client-rpc.js:79]: It might involve more network round trips but you could also change it to _always_ get a fresh grant token from the backend whenever the client requests one.

File: lms/views/api/grant_token.py (hunk 3 above)
Comment (seanh) [reply lms/views/api/grant_token.py:12]: What is the `context` object during one of these `/api/grant_token` requests?

For example during an LTI launch request the context is an [`LTILaunchResource`](https://github.com/hypothesis/lms/blob/master/lms/resources/lti_launch.py). There's also [`OAuth2RedirectResource`](https://github.com/hypothesis/lms/blob/master/lms/resources/oauth2_redirect.py) for OAuth 2 redirect requests.

I think you may want to add a `GrantTokenResource` class for this `/grant_token` route in `lms/resources/`. That resource class can then have a `context.grant_token` property that would call the code that's currently in `JSConfig._grant_token()` (that code would move to a private helper function within `/lms/resources/`, called by both `JSConfig` and the new resource class).

**Comment by seanh** [comment]
How is the frontend's request to the new `/api/grant_token` endpoint authenticated? Is it authenticated using the JWT that the frontend usually uses to authenticate its requests to the backend? If so, now long does this JWT last until it expires? Might you have the same problem if, by the time the user clicks <kbd>Open Notebook</kbd>, the JWT has already expired?

You might be able to fix this by increasing the lifetime of the JWT (which is probably a good idea anyway: if it's too short I think it might expire during grading and cause LMS app breakage).

But if the JWT has a limited lifetime (even if we've increased that lifetime long enough that we think it's okay) I'm asking myself: why implement the `/api/grant_token` endpoint? Why not just increase the lifetime of the grant token as well?

**Comment by robertknight** [comment]
Thanks for the feedback. I've done some further work on this today and I think it will make sense to split into separate changes: One PR to extract the existing grant-token generation logic out of `JSConfig` so it can be re-used elsewhere, and another to implement the grant token refresh.
